Plot Histograms and Specified ODB's

DLBPointon opened the pull request:
Adding support for genescopefk plots

- NOT TESTED due to error with ncbi datasets.

DLBPointon committed:
Added some logic - needs testing

DLBPointon committed:
Starting plot_histo

DLBPointon committed:
Update to add in the PLOT_HISTOGRAM subworkflow

DLBPointon committed:
Update to add in the PLOT_HISTOGRAM subworkflow

DLBPointon committed:
Update to add in the PLOT_HISTOGRAM subworkflow

muffato commented on CHANGELOG.md:
```diff
@@ -3,6 +3,12 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [[2.1.2](https://github.com/sanger-tol/genomenote/releases/tag/2.1.2)] - Pembroke Welsh Corgi (patch 2) [2025-03-25]
```
It will be v2.2, not v2.1.2, since we're adding features

DLBPointon replied:
Done

DLBPointon replied:
Done

muffato commented on CHANGELOG.md:
```diff
@@ -3,6 +3,12 @@
+
+### Enhancements & fixes
+
+- Addition of the GFASTATS module and config.
```
List the two other changes you're making

DLBPointon replied:
Done

DLBPointon replied:
Done

muffato commented on subworkflows/local/plot_histogram.nf:
```diff
@@ -0,0 +1,53 @@
+    //
+    // MODULE: FASTK to calculate hist data for genescope
+    //
+    FASTK_FASTK( CAT_CAT.out.outfile )
```
FASTK_FASTK is already run in GENOME_STATISTICS. I would rather add the HISTEX and GENESCOPE modules over there right net to it.

DLBPointon replied:
Re-implementing now

muffato commented on subworkflows/local/genome_statistics.nf:
```diff
@@ -36,17 +36,25 @@ workflow GENOME_STATISTICS {
+    // LOGIC: If the obd param is manually specified then we use that,
+    //          otherwise we get the lineage from NCBI.
+    //
+    if (!params.odb) {}
```
Remove the closing curly bracket.

Also, it's preferred to have positive conditions in the `if`, i.e. `if (params.odb)`

DLBPointon replied:
That shouldn't have been in this PR i'm removing it for now

muffato reviewed (commented):
Will need to add the new Busco parameter in `nextflow.config`, the schema, etc.

Can you try running the pipeline locally in the meantime ? The only parameters needed are `-profile test,sanger,singularity` and `--outdir`
There are several syntax errors in this PR that prevent the pipeline from even starting
PLOT_HISTOGRAM

DLBPointon commented:
I will test once back at the desk,

TBF the specified_odb changes were supposed to be part of their own pr. But I can keep them here if that's ok?

muffato commented:
> TBF the specified_odb changes were supposed to be part of their own pr. But I can keep them here if that's ok?

Let's keep the change here. Please update the title/description of the PR

DLBPointon committed:
Updates - not complete

DLBPointon committed:
Updated CHANGELOG

DLBPointon committed:
Updated to reflect comments

DLBPointon committed:
Re-adding specified_odb with args in config too

DLBPointon committed:
Forgot to commit histex

DLBPointon commented:
I've opened this: https://github.com/ncbi/datasets/issues/468

Locally, much of the pipeline works. To test more I will comment out some processes and get that test coverage up.

DLBPointon commented:
Histograms work until:
```
Command error:
  Error in scan(file = file, what = what, sep = sep, quote = quote, dec = dec,  : 
    scan() expected 'a real', got 'Histogram'
  Calls: read.csv -> read.table -> scan
  Execution halted
  grep: *_model.txt: No such file or directory
```

DLBPointon committed:
Updating the FASTK>GENOMESCOPE work

DLBPointon committed:
Uncommenting everything

DLBPointon committed:
Module def was uncommented by accident

DLBPointon committed:
Updating kmer to better generic option

DLBPointon commented:
I fixed the above error,

After speaking to NCBI, the IP issue is fixed too. 

Locally, it runs fine. the only issue is the Actions runner:
```
.command.sh: line 7:    32 Killed                  GeneScopeFK.R -k31 --input ilCerPisi1.hist --output . --name_prefix ilCerPisi1
  grep: *_model.txt: No such file or directory
```

muffato commented on subworkflows/local/genome_statistics.nf:
```diff
@@ -23,37 +26,49 @@ workflow GENOME_STATISTICS {
-    | splitCsv()
-    | map { row -> row[1] }
-    | set { ch_lineage }
+        ch_lineage.view{"AUTO_LINEAGE: $it"}
```
No view() in the final code

DLBPointon replied:
Just for testing so i could see the output from this logic.

DLBPointon replied:
Removed now

muffato commented on conf/modules.config:
```diff
@@ -19,6 +19,14 @@ process {
         ]
     }
 
+    withName: GENESCOPEFK {
+        publishDir = [
+            path: { "${params.outdir}/pipeline_info/genomenote/genomescope" },
+            mode: params.publish_dir_mode,
+            saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
+        ]
+    }
+
```
Merge into the other `withName: GENESCOPEFK`

muffato replied:
Also: not sure why but in my tests those files were not copied to the results directory

DLBPointon replied:
All fixed,

Did you check in the pipeline_info folder? Accidently sent them there instead of main results folder

muffato commented on CHANGELOG.md:
```diff
@@ -3,6 +3,33 @@
+
+| Dependency     | Old version | New version |
+| -------------- | ----------- | ----------- |
+| `gfastats`     |             |             |
+| `cat`          |             | 2.3.4       |
+| `genescopefk`  |             | 1.2         |
+| `fastk_histex` |             |             |
```
`gfastats` and `fastk_histex` need a version number too

DLBPointon replied:
Added

muffato commented on CHANGELOG.md:
```diff
@@ -3,6 +3,33 @@
+
+| Old parameter | New parameter  |
+| ------------- | -------------- |
+|               | --odb_override |
```
I'd suggest naming the option `busco_lineage`, similar to `busco_lineages` in BTK.

DLBPointon replied:
Done

muffato commented on conf/base.config:
```diff
@@ -94,9 +94,14 @@ process {
         cpus   = { log_increase_cpus(2, 2*task.attempt, 1, 2) }
         time   = { check_max( 2.h * task.attempt, 'time') }
     }
+
+    withName: GFASTATS {
+        memory = { check_max( 1.GB * 1.2 * Math.ceil(meta.genome_size/1e9) * task.attempt, 'memory' ) }
+    }
```
Oh you added it here ! I added it in #176 too

muffato replied:
We need memory requirements for FASTK_HISTEX and GENESCOPEFK

DLBPointon replied:
Probably as part of me mixing up my PR's

Added a basic mem requirement for now.

muffato commented on conf/test.config:
```diff
@@ -15,39 +15,39 @@ params {
 
     // Reducing the k-mer size to speed FastK/Merqury a little bit, but also decrease the memory consumption
-    kmer_size = 11
+    kmer_size                   = 31
```
Why do you increase the k-mer size her ? We set it low to make the tests run faster / with less memory.

I bet it's what's causing the CI to fail
```
.command.sh: line 7:    32 Killed                  GeneScopeFK.R -k31 --input ilCerPisi1.hist --output . --name_prefix ilCerPisi1
  grep: *_model.txt: No such file or directory
```

DLBPointon replied:
I was using this for testing a real value on Actions, testing with kmer of 11 returned the same error so i thought i'd test it with a value we are more likely to use. Odd that it still failed, not sure what is happening there when it runs locally fine.

I'll change it back so it isn't left like this, but i did test on 11 first.

muffato commented on nextflow.config:
```diff
@@ -238,12 +239,12 @@ dag {
 manifest {
     name            = 'sanger-tol/genomenote'
-    author          = """@cibinsb, @muffato, @gq1, @priyanka-surana, @BethYates"""
+    author          = """@cibinsb, @muffato, @gq1, @priyanka-surana, @BethYates, @DLBPointon"""
```
Feel free to update CITATION.cff too

DLBPointon replied:
Will do

muffato commented on nextflow.config:
```diff
@@ -238,12 +239,12 @@ dag {
     mainScript      = 'main.nf'
     nextflowVersion = '!>=22.10.1'
-    version         = '2.1.1'
+    version         = '2.1.2'
```
```suggestion
    version         = '2.2.0'
```

DLBPointon replied:
Done

muffato reviewed (commented):
Thank you for pushing this, Damon ! See my review below

DLBPointon committed:
Fixing suggestions from PR and adding citations

DLBPointon committed:
Fixing suggestions from PR and adding citations

DLBPointon committed:
Removed sums for mem... they don't need enough for that

DLBPointon commented:
> Thank you for pushing this, Damon ! See my review below

Thanks @muffato , I've addressed all the comments and as a bonus the tests are now passing (adding the memory allocations helped there) ... a couple of fails due to a gateway error with cooler dump, but a re-run is passing now.

DLBPointon committed:
Updated the resource limits to the allowed max - FREE!

DLBPointon commented:
Once again: `Process exceeded running time limit (1h)`

DLBPointon committed:
Merge branch 'dev' into dp24_plot_histos

DLBPointon committed:
Updating some errors that appeared in modules.json

DLBPointon commented:
@muffato 
This is going to end up the same state of the previous PR. 1 test run passes but the other doesn't solely because of MultiQC being a pain. Everything else looks good though.

muffato commented:
That's OK, I'll review the branch as it is. I've created #178 to track and solve the MultiQC issue

muffato committed:
Raising the memory of MultiQC

It helps on LSF

muffato committed:
Harshil alignment ftw !

muffato committed:
Whitespace/line editing to make the two files more similar

muffato committed:
Validate the busco parameter

muffato committed:
Removed duplicated GFASTATS entry and moved the *FK rules together

muffato committed:
Version bump

muffato committed:
Moved to the alphabetical order

muffato committed:
The date will be provided later

muffato committed:
Avoid mentioning odb

muffato committed:
Not needed any more

muffato commented on nextflow_schema.json:
```diff
@@ -48,6 +48,12 @@
+                "busco_lineage": {
+                    "type": "string",
+                    "description": "Override the BUSCO database used for the assembly assessment.",
+                    "pattern": "._odb10$",
```
@DLBPointon . I've added this and removed your comment about validating the busco lineage in PIPELINE_INITIALISATION

muffato committed:
[lint] there can't be a default null value with a pattern
